### ISSUE TITLE
Not working with classes with lambdas (retrolambda)

### ISSUE THREAD

**Issue by SammyVimes** [robolectric/robolectric#2364]
When you try to shadow class, which has lambdas inside, its creationfails with 
`NoClassDefFoundError` with cause:
`java.lang.ClassCastException: org.objectweb.asm.tree.InvokeDynamicInsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode
at org.robolectric.internal.bytecode.InstrumentingClassLoader$ClassInstrumentor.filterSpecialMethods(InstrumentingClassLoader.java:683)`

**Comment by karlicoss** [comment]
Just curious: how is it possible to have 'real' lambdas in an Android class? What are you trying to shadow?

**Comment by SammyVimes** [comment]
I forgot to mention, that I use retrolambda :D

**Comment by jaredsburrows** [comment]
I use JDK8 + Retrolambda just fine. Please show some example code.

**Comment by SammyVimes** [comment]
@jaredsburrows 

```
class Taskhandler {
    [...]
    public void onHandleOperationResult(final Object result,
                                            final String error,
                                            final OperationRunner operationRunner,
                                            final Operation operation) {
            mainHandler.post(() -> {
                [....]
            });
}
```
```
@Implements(TaskHandler.class)
public class TaskHandlerShadow {
    [...]
    @Implementation
    public void onHandleOperationResult(final Object result,
                                        final String error,
                                        final OperationRunner operationRunner,
                                        final Operation operation) { 
        [...]
    }
}
```

And then constructor `new TaskHandler()` is called I got 
```
java.lang.NoClassDefFoundError ....TaskHandler
...
Caused by: java.lang.ClassNotFoundException: couldn't load ....TaskHandler
...
Caused by: java.lang.ClassCastException: org.objectweb.asm.tree.InvokeDynamicInsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode
```

**Comment by jaredsburrows** [comment]
Usually when people use `Retrolambda`, they use it with `RxJava`. Why are you creating custom handlers? Maybe your Shadow is wrong? Shadows should be last resort.

**Comment by SammyVimes** [comment]
@jaredsburrows 
Usually there should not be ClassCastExceptions, no matter what library is used or not used :D. It certainly is a bug. 
About RxJava: I simply don't have enough experience in it to use it in production, so I wrote some custom classes, which are easy as a brick.

**Comment by jongerrish** [comment]
Do you really need to shadow your own classes?

I recommend using a mocking library instead for that.

I do think you're right though, this could be a bug related to the use of
intrinsics that were added recently. I believe you can switch this off with
a system property though.
On Mar 21, 2016 10:20 AM, "Semyon Danilov" <notifications@github.com> wrote:

@jaredsburrows <https://github.com/jaredsburrows>
Usually there should not be ClassCastExceptions, no matter what library is
used or not used :D. It certainly is a bug.
About RxJava: I simply don't have enough experience in it to use it in
production, so I wrote some custom classes, which are easy as a brick.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/robolectric/robolectric/issues/2364#issuecomment-199385216>

**Comment by SammyVimes** [comment]
@jongerrish 
I am not sure if I should shadow them. I considered using PowerMock's `supress`, but it seems to be Shadow's analogue.
This class's (that I shadow) object, is instantiating deep inside in app's logic. I actually think of something like factory injecting, but it's sure is not so easy to implement in Android (comparing with Spring etc).

**Comment by budnyjj** [comment]
I've just encountered the same issue, trying to shadow a class, which uses retrolambdas.
So I am getting 
java.lang.NoClassDefFoundError
caused by java.lang.ClassNotFoundException
caused by java.lang.ClassCastException: org.objectweb.asm.tree.InvokeDynamicInsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode.
It happens only when I try to run tests from Android Studio; test are passed when I run them from console.

**Comment by jaredsburrows** [comment]
@budnyjj Can you show an example test? I use retrolambda in my apps just fine.

**Comment by budnyjj** [comment]
Engine.java:
```java
public class Engine {
    public static final Callback EMPTY_CALLBACK = args -> {
    };
}
```

EngineShadow.java:
```java
@Implements(Engine.class)
public class EngineShadow {

}
```

CustomShadowTestRunner.java:
```java
public class CustomShadowTestRunner extends RobolectricGradleTestRunner {

    public CustomShadowTestRunner(Class<?> testClass) throws InitializationError {
        super(testClass);
    }

    @Override
    public InstrumentationConfiguration createClassLoaderConfig() {
        return InstrumentationConfiguration.newBuilder()
                .addInstrumentedClass(Engine.class.getName())
                .build();
    }
```

I'm runnign my tests on Linux as well as on Mac using robolectric 3.0 with retrolambda plugin 3.2.5.

**Comment by jaredsburrows** [comment]
So this should work normally. Are you only seeing this error when using shadows + retrolambda?

**Comment by budnyjj** [comment]
And the test is:
```java
@Config(shadows = {EngineShadow.class})
@RunWith(CustomShadowTestRunner.class)
public class JSEngineTest {
    // TODO: add unit tests
    @Test
    public void testBasic() {
        Assert.assertTrue(true);
    }
}
```

**Comment by budnyjj** [comment]
Yes, each time I replace 
```java
public static final Callback EMPTY_CALLBACK = args -> {
    };
```
with
```java
    public static final Callback EMPTY_CALLBACK = new Callback() {
        @Override
        public void invoke(String[] args) {
            
        }
    }
```
it passes.

**Comment by jaredsburrows** [comment]
@SammyVimes @budnyjj This seems to work fine for me. I am using Robolectric 3.1.1.

`gradle clean testDebug --tests="*JSEngineTest*"`

CustomShadowTestRunner.java
```
import org.junit.runners.model.InitializationError;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;
import org.robolectric.internal.bytecode.InstrumentationConfiguration;

public class CustomShadowTestRunner extends RobolectricTestRunner {

    public CustomShadowTestRunner(Class<?> testClass) throws InitializationError {
        super(testClass);
    }

    @Override
    public InstrumentationConfiguration createClassLoaderConfig(Config config) {
        super.createClassLoaderConfig(config);

        return InstrumentationConfiguration.newBuilder()
                .addInstrumentedClass(Engine.class.getName())
                .build();
    }
}
```

Engine.java
```
public class Engine {

    interface Callback {
        void invoke(String[] args);
    }

    public static final Callback EMPTY_CALLBACK = args -> {
    };

}
```

EngineShadow.java
```
import org.robolectric.annotation.Implements;

@Implements(Engine.class)
public class EngineShadow {

}

```

JSEngineTest.java
```
import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.annotation.Config;

@Config(shadows = {EngineShadow.class})
@RunWith(CustomShadowTestRunner.class)
public class JSEngineTest {

    @Test
    public void testBasic() {
        Assert.assertTrue(true);
    }
}
```

**Comment by jaredsburrows** [comment]
@SammyVimes Can we close this? @jongerrish @xian

**Comment by budnyjj** [comment]
Sorry, my mistake. Here is failing test body:
```java
@Config(constants = BuildConfig.class, shadows = {EngineShadow.class}, sdk = 21)
@RunWith(CustomShadowTestRunner.class)
public class ExampleUnitTest {
    @Test
    public void testBasic() {
        Engine.EMPTY_CALLBACK.invoke(new String[]{"abc"});
        assertTrue(true);
    }
}
```
Error message:
```java
java.lang.NoClassDefFoundError: budnyjj/robolambda/Engine
	at budnyjj.robolambda.ExampleUnitTest.testBasic(ExampleUnitTest.java:14)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:251)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:188)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:152)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: java.lang.ClassNotFoundException: couldn't load budnyjj.robolambda.Engine
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.findClass(InstrumentingClassLoader.java:155)
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.loadClass(InstrumentingClassLoader.java:95)
	... 28 more
Caused by: java.lang.ClassCastException: org.objectweb.asm.tree.InvokeDynamicInsnNode cannot be cast to org.objectweb.asm.tree.MethodInsnNode
	at org.robolectric.internal.bytecode.InstrumentingClassLoader$ClassInstrumentor.filterNasties(InstrumentingClassLoader.java:626)
	at org.robolectric.internal.bytecode.InstrumentingClassLoader$ClassInstrumentor.instrument(InstrumentingClassLoader.java:386)
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.getInstrumentedBytes(InstrumentingClassLoader.java:243)
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.findClass(InstrumentingClassLoader.java:148)
	... 29 more
```

**Comment by jaredsburrows** [comment]
@budnyjj Now that fails for me. This is just hard for Robolectric to handle since Retrolambda takes Java8 byte code and turn it into Java7 byte code.

**Comment by budnyjj** [comment]
What are the options to do in this situation?
It turns that Robolectric and Retrolambda are not compatible with each other in general, right?

**Comment by jaredsburrows** [comment]
I guess. What is the use that the class you are using needs a shadow? or maybe for now, do not shadow classes with jdk8 code or just switch those lambdas to a jdk7 implementation.

**Comment by budnyjj** [comment]
This shadow class should mock the original one, which has a number of native methods.

**Comment by jaredsburrows** [comment]
You can just try mocking it for now or use a java 7 implementation instead of lambdas for the class that you are shadowing.

**Comment by budnyjj** [comment]
Yes, I'll try to mock them.
Thank you!

**Comment by prattpratt** [comment]
I had the same issue. But I'm using plain IDEA for development rather than AS. And I noticed that if I use JUnit Runner (IDE internals configured in Settings) then such test really fails but if I choose Gradle Runner - such tests passes. So my conclusion is that when IDE runs tests using JUnit Runner then we get Java 8 code with real lambdas rather then retrolambdas. And seems that roboelectric does not support Java 8. I looked in the source code where stacktrace points to. And there's no special handling for invokedynamic opcode. It's the same as for all other invokeXXX methods.
